clang/lib/Sema/SemaChecking.cpp

```diff
@@ -5553,6 +5557,58 @@ bool Sema::BuiltinSetjmp(CallExpr *TheCall) {
   return false;
 }
 
+bool Sema::BuiltinCountedByRef(CallExpr *TheCall) {
+  // For simplicity, we support only a limited expressions for the argument.
+  // Specifically 'ptr->array' and '&ptr->array[0]'. This allows us to reject
+  // arguments with complex casting, which really shouldn't be a huge problem.
+  if (checkArgCount(TheCall, 1))
+    return true;
+
+  ExprResult ArgRes = UsualUnaryConversions(TheCall->getArg(0));
+  if (ArgRes.isInvalid())
+    return true;
+
+  const Expr *Arg = ArgRes.get()->IgnoreParenImpCasts();
+  if (!isa<PointerType>(Arg->getType()) && !Arg->getType()->isArrayType())
```

Comment:
How do you get an array type after the usual unary conversions? That seems like it would have already applied the array decay, no?

Reply:
```c
struct x {
  int count;
  char *array[];
} *p;

p->array; // << is an array type.
```

Reply:
I think `p->array` is a pointer type because of array decay: https://godbolt.org/z/aY4T18sdE
```
|   `-ImplicitCastExpr <line:7:3, col:6> 'char **' <ArrayToPointerDecay>
|     `-MemberExpr <col:3, col:6> 'char *[]' lvalue ->array 0xd965fb8
|       `-ImplicitCastExpr <col:3> 'struct x *' <LValueToRValue>
|         `-DeclRefExpr <col:3> 'struct x *' lvalue Var 0xd9660d8 'p' 'struct x *'
```

Reply:
Notice the implicit cast expr. Those are removed by the `const Expr *Arg` line above it.

Reply:
Oh gosh, yeah, this is my mistake, sorry for the noise @bwendling!